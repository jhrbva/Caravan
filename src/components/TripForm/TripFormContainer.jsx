import React from 'react';
import axios from 'axios';
import { Formik } from 'formik';
import { withRouter } from 'react-router';

import Navbar from '../Navbar/Navbar';
import BigButton from '../BigButton/BigButton';

export class TripFormContainer extends React.Component {
	static Page = ({ children }) => children;

	constructor(props) {
		super(props);
		this.state = {
			page: 0,
			values: props.initialValues,
		};
	}

	nextPage = (values) =>
		this.setState((state) => ({
			page: Math.min(state.page + 1, this.props.children.length - 1),
			values,
		}));

	previousPage = () =>
		this.setState((state) => ({
			page: Math.max(state.page - 1, 0),
		}));

	pgFormatDate = (date) => {
		return date.toISOString().replace('T', ' ').replace('Z', '');
	};

	onSubmit = (values, bag) => {
		const { history } = this.props;
		const {
			start_location,
			destination,
			start_date,
			trip_description,
			trip_title,
			guests,
		} = values;
		//console.log('in onSubmit', values, bag);

		axios
			.post('/trip', {
				host_id: 16,
				start_location,
				destination,
				trip_date: this.pgFormatDate(start_date),
				trip_description,
				trip_title,
			})
			.then(function (response) {
				fetch(`/user/${guests}`)
					.then((res) => {
						return res.json();
					})
					.then((info) => {
						axios.post('/invitations', {
							host_id: 16,
							user_id: info[0].userid,
							trip_id: response.data,
						});
					});
				history.push('/dashboard');
			})
			.catch(function (error) {
				console.log(error);
			});
	};

	handleSubmit = (values, bag) => {
		const { children } = this.props;
		const { page } = this.state;
		const isLastPage = page === React.Children.count(children) - 1;
		if (isLastPage) {
			return this.onSubmit(values, bag);
		} else {
			bag.setTouched({});
			bag.setSubmitting(false);
			this.nextPage(values);
		}
	};

	render() {
		const { children } = this.props;
		const { page, values } = this.state;
		const activePage = React.Children.toArray(children)[page];
		const isLastPage = page === React.Children.count(children) - 1;

		return (
			<>
				<Navbar />
				<Formik
					initialValues={values}
					enableReinitialize={false}
					onSubmit={this.handleSubmit}
					render={({ values, handleSubmit, isSubmitting, handleReset }) => (
						<form onSubmit={handleSubmit}>
							{activePage}
							<div className='buttons'>
								{page > 0 && (
									<BigButton
										type='button'
										onClick={this.previousPage}
										value='Previous'
									/>
								)}

								{!isLastPage && <BigButton type='submit' value='Next' />}
								{isLastPage && (
									<BigButton
										type='submit'
										value='Submit'
										disabled={isSubmitting}
									/>
								)}
							</div>
						</form>
					)}
				/>
			</>
		);
	}
}

export default withRouter(TripFormContainer);
