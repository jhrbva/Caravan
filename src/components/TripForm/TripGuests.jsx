import React from 'react';
import { Field, Form, Formik } from 'formik';

import '../../utilities/fonts.scss';
import Input from '../Input/Input';
import BigButton from '../BigButton/BigButton';
import {
	required,
} from '../../utilities/formValidation';


const TripGuests = () => (
	<Form>
		<Field
			icon='group_add'
			type='text'
			name='guests'
			placeholder='Guest'
			validate={required}
			autofocus='true'
			component={Input}
		/>
		<BigButton value={'Back'} />
		<BigButton value={'Next'} />
	</Form>
)

export default TripGuests;
