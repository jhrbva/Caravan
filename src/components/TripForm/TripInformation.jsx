import React from 'react';
import { Field } from 'formik';
import '../../utilities/fonts.scss';
import Input from '../Input/Input';
import { required } from '../../utilities/formValidation';

import TitleIcon from '@material-ui/icons/Title';
import DescriptionIcon from '@material-ui/icons/Description';
import TripOriginIcon from '@material-ui/icons/TripOrigin';
import LocationCityIcon from '@material-ui/icons/LocationCity';
import EventIcon from '@material-ui/icons/Event';
import ScheduleIcon from '@material-ui/icons/Schedule';

const TripInformation = () => (
	<>
		<Field
			icon={<TitleIcon />}
			type='text'
			name='trip_title'
			placeholder='Trip Title'
			validate={required}
			autoFocus={true}
			component={Input}
		/>
		<Field
			icon={<DescriptionIcon />}
			type='text'
			name='trip_description'
			placeholder='Description'
			validate={required}
			autoFocus={true}
			component={Input}
		/>
		<Field
			icon={<TripOriginIcon />}
			type='text'
			name='start_location'
			placeholder='Start Location'
			validate={required}
			autoFocus={true}
			component={Input}
		/>
		<Field
			icon={<LocationCityIcon />}
			type='text'
			name='destination'
			placeholder='Destination'
			validate={required}
			autoFocus={false}
			component={Input}
		/>
		<Field
			icon={<EventIcon />}
			type='text'
			name='start_date'
			placeholder='Date'
			validate={required}
			autoFocus={false}
			component={Input}
		/>
		<Field
			icon={<ScheduleIcon />}
			type='text'
			name='start_time'
			placeholder='Time'
			validate={required}
			autoFocus={false}
			component={Input}
		/>
	</>
);

export default TripInformation;
