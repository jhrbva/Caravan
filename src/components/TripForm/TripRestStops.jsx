import React from 'react';
import { Field, Form } from 'formik';
import '../../utilities/fonts.scss';
import Input from '../Input/Input';

const TripRestStops = () => (
	<Form>
		<Field
			icon='location_on'
			type='text'
			name='rest_stops'
			placeholder='Rest Stop'
			autoFocus={true}
			component={Input}
		/>
	</Form>
);

export default TripRestStops;
