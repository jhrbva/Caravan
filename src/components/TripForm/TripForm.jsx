import React from 'react';
import { Formik } from 'formik';

import BigButton from '../BigButton/BigButton';
import TripInformation from './TripInformation';
import TripGuests from './TripGuests';
import TripRestStops from './TripRestStops';

class TripContainer extends React.Component {
	static Page = ({ children }) => children;

	constructor(props) {
		super(props);
		this.state = {
			page: 0,
			values: props.initialValues,
		};
	}

	next = values =>
		this.setState(state => ({
			page: Math.min(state.page + 1, this.props.children.length - 1),
			values,
		}));

	previous = () =>
		this.setState(state => ({
			page: Math.max(state.page - 1, 0),
		}));

	validate = values => {
		const activePage = React.Children.toArray(this.props.children)[
			this.state.page
		];
		return activePage.props.validate ? activePage.props.validate(values) : {};
	};

	handleSubmit = (values, bag) => {
		const { children, onSubmit } = this.props;
		const { page } = this.state;
		const isLastPage = page === React.Children.count(children) - 1;
		if (isLastPage) {
			return onSubmit(values, bag);
		} else {
			bag.setTouched({});
			bag.setSubmitting(false);
			this.next(values);
		}
	};

	render() {
		const { children } = this.props;
		const { page, values } = this.state;
		const activePage = React.Children.toArray(children)[page];
		const isLastPage = page === React.Children.count(children) - 1;
		return (
			<Formik
				initialValues={values}
				enableReinitialize={false}
				validate={this.validate}
				onSubmit={this.handleSubmit}
				render={({ values, handleSubmit, isSubmitting, handleReset }) => (
					<form onSubmit={handleSubmit}>
						{activePage}
						<div className='buttons'>
							{page > 0 && (
								<BigButton
									type='button'
									onClick={this.previous}
									value='Previous'
								/>
							)}

							{!isLastPage && <BigButton type='submit' value='Next' />}
							{isLastPage && (
								<BigButton
									type='submit'
									value='Submit'
									disabled={isSubmitting}
								/>
							)}
						</div>
					</form>
				)}
			/>
		);
	}
}

const TripForm = () => (
	<>
		<TripContainer
			initialValues={{
				start_location: '',
				destination: '',
				start_date: '',
				start_time: '',
				guests: '',
				rest_stops: '',
			}}
		>
			<TripContainer.Page>
				<TripInformation />
			</TripContainer.Page>
			<TripContainer.Page>
				<TripGuests />
			</TripContainer.Page>
			<TripContainer.Page>
				<TripRestStops />
			</TripContainer.Page>
		</TripContainer>
	</>
);

export default TripForm;
