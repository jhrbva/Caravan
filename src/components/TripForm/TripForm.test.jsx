import React from 'react';
import { shallow } from 'enzyme';
import TripForm, { TripContainer } from './TripForm';

describe('TripForm', () => {
	it('should match the snapshot', () => {
		const wrapper = shallow(<TripForm />);
		expect(wrapper).toMatchSnapshot();
	});
});
describe('NextPage()', () => {
	it('changes pages from 1 to 2 then 3', () => {
		const wrapper = shallow(<TripContainer children={[1, 2, 3]} />);
		const instance = wrapper.instance();
		expect(instance.state.page).toBe(0);
		instance.nextPage({ preventDefault: jest.fn() });
		expect(instance.state.page).toBe(1);
		instance.nextPage({ preventDefault: jest.fn() });
		expect(instance.state.page).toBe(2);
	});
});
describe('PreviousPage()', () => {
	it('changes pages from 3 to 2 then 1', () => {
		const wrapper = shallow(<TripContainer />);
		const instance = wrapper.instance();
		wrapper.setState({ page: 3 });
		instance.previousPage({ preventDefault: jest.fn() });
		expect(instance.state.page).toBe(2);
		instance.previousPage({ preventDefault: jest.fn() });
		expect(instance.state.page).toBe(1);
	});
});
