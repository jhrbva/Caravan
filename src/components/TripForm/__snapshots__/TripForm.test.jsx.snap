// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripForm should match the snapshot 1`] = `
<Fragment>
  <TripContainer
    initialValues={
      Object {
        "destination": "",
        "guests": "",
        "rest_stops": "",
        "start_date": "",
        "start_location": "",
        "start_time": "",
      }
    }
  >
    <Component>
      <TripInformation />
    </Component>
    <Component>
      <TripGuests />
    </Component>
    <Component>
      <TripRestStops />
    </Component>
  </TripContainer>
</Fragment>
`;
